File: backend/app/schemas/booking_emergency_contact.py
"""预订紧急联系人校验 schema。"""
import re
from datetime import date

from pydantic import BaseModel, field_validator


class BookingEmergencyContactValidation(BaseModel):
    """紧急联系人校验请求。"""
    chinese_name: str | None = None
    relationship: str | None = None
    given_name_pinyin: str | None = None
    surname_pinyin: str | None = None
    relation: str | None = None
    birth_date: date | None = None
    phone: str | None = None
    email: str | None = None
    gender: str | None = None
    region: str | None = None
    address_detail: str | None = None
    postal_code: str | None = None
    consultant_id: str | None = None
    address_line: str | None = None
    phone_country_code: str | None = None
    country_code: str | None = None
    country_name: str | None = None
    level1_code: str | None = None
    level1_name: str | None = None
    city_code: str | None = None
    city_name: str | None = None
    district_code: str | None = None
    district_name: str | None = None
    same_as_personal_address: bool | None = None

    @field_validator("surname_pinyin", "given_name_pinyin")
    @classmethod
    def normalize_pinyin(cls, value: str | None) -> str | None:
        if value is None:
            return value
        normalized = value.upper()
        if not re.fullmatch(r"[A-Z -]+", normalized):
            raise ValueError("拼音仅支持英文大写字母、空格和连字符")
        return normalized

    @field_validator("birth_date")
    @classmethod
    def validate_age(cls, value: date | None) -> date | None:
        if value is None:
            return value
        today = date.today()
        age = today.year - value.year - ((today.month, today.day) < (value.month, value.day))
        if age < 18:
            raise ValueError("紧急联系人须年满 18 周岁")
        if age > 100:
            raise ValueError("紧急联系人年龄不能超过 100 周岁")
        return value


class BookingEmergencyContactValidationRead(BaseModel):
    valid: bool = True

File: frontend/src/stores/bookingEmergencyContact.ts
/** 预订流程紧急联系人 Store — 跨步骤保持表单状态，支持从个人信息复制地址。 */
import { defineStore } from 'pinia'
import { reactive, ref } from 'vue'
import type { EmergencyContactForm } from '@/utils/emergencyContactValidation'
import { useBookingPersonalInfoStore } from '@/stores/bookingPersonalInfo'

function createEmptyForm(): EmergencyContactForm {
  return {
    chinese_name: '',
    given_name_pinyin: '',
    surname_pinyin: '',
    relationship: '',
    birth_date: '',
    phone: '',
    phone_country_code: '+86',
    email: '',
    gender: '',
    region: '',
    address_line: '',
    postal_code: '',
    consultant_id: '',
    country_code: '',
    country_name: '',
    level1_code: '',
    level1_name: '',
    city_code: '',
    city_name: '',
    district_code: '',
    district_name: '',
    address_detail: '',
  }
}

export const useBookingEmergencyContactStore = defineStore('bookingEmergencyContact', () => {
  const form = reactive<EmergencyContactForm>(createEmptyForm())
  const sameAsApplicant = ref(false)

  function reset() {
    Object.assign(form, createEmptyForm())
    sameAsApplicant.value = false
  }

  /** 将地址字段切换为与申请人相同或清空。 */
  function toggleSameAddress(enabled: boolean) {
    sameAsApplicant.value = enabled
    if (enabled) {
      const applicant = useBookingPersonalInfoStore()
      copyApplicantAddress(applicant.form)
    }
  }

  function copyApplicantAddress(applicantForm: Record<string, any>) {
    const addressFields = [
      'country_code', 'country_name', 'level1_code', 'level1_name',
      'city_code', 'city_name', 'district_code', 'district_name', 'region',
      'postal_code',
    ] as const
    addressFields.forEach((field) => { form[field] = applicantForm[field] || '' })
    form.address_line = applicantForm.address_line || applicantForm.address_detail || ''
    form.address_detail = form.address_line
  }

  /** 别名，兼容旧 API：从申请人数据复制地址字段。 */
  function setSameAsApplicant(enabled: boolean, applicantForm?: Record<string, any>) {
    sameAsApplicant.value = enabled
    if (enabled && applicantForm) {
      copyApplicantAddress(applicantForm)
    }
  }

  return { form, sameAsApplicant, reset, toggleSameAddress, setSameAsApplicant }
})
